### PR TITLE
NATIVE-500: Feedback fails with 400

### PR DESCRIPTION
**Pull request by steffenkleinle**
This PR does not fix the bug, because I could not reproduce it. Instead, I added more information to the ResponseError, which is thrown if something different than status code 200 is returned. This should hopefully help us to identify the issue.

### CODE REVIEW

File: src/Endpoint.js [hunk 1]
```diff
@@ -60,15 +59,11 @@ class Endpoint<P, T> {
...
-      response = await (this.mapParamsToBody ? this.postFormData(url, params) : fetch(url))
-    } catch (e) {
-      throw new FetchError({ endpointName: this.stateName, innerError: e })
-    }
+    const formData = this.mapParamsToBody ? this.mapParamsToBody(params) : null
+    const response = await this.fetchOrPost(url, formData)
 
     if (!response.ok) {
-      throw new ResponseError({ endpointName: this.stateName, response })
+      throw new ResponseError({ endpointName: this.stateName, responseStatus: response.status, url, formData })
```
Comment (michael-markl) [inline src/Endpoint.js]: Was having `response` in the error not helpful? I guess we could extract the server's response (i.e. in text form) from that. Or is there another reason you replaced it with response.status?

File: src/Endpoint.js [hunk 2]
```diff
@@ -60,15 +59,11 @@ class Endpoint<P, T> {
       return new Payload(false, url, this.responseOverride, null)
     }
 
-    let response = null
-    try {
-      response = await (this.mapParamsToBody ? this.postFormData(url, params) : fetch(url))
-    } catch (e) {
-      throw new FetchError({ endpointName: this.stateName, innerError: e })
```
Comment (maxammann) [inline src/Endpoint.js:67]: Why did you remove FetchError here? Is it used anywhere now?

File: src/Endpoint.js [hunk 3]
```diff
@@ -38,16 +38,15 @@ class Endpoint<P, T> {
     return this._stateName
   }
 
-  async postFormData (url: string, params: P): Promise<Response> {
-    if (!this.mapParamsToBody) {
-      throw new Error(`The endpoint ${this.stateName} is not able to post form data!`)
+  async fetchOrPost (url: string, formData: ?FormData): Promise<Response> {
```
Comment (maxammann) [inline src/Endpoint.js]: I don't really understand this name as this always does a `fetch` which uses the POST method.

Comment (maxammann) [reply src/Endpoint.js]: Ah oke now I saw the trinary.

In that case this function should be called `getOrPost`. I'd prefere how it was before this as a GET and a POST usually requires different parameters.

**Review by maxammann — CHANGES_REQUESTED** [review]

File: src/Endpoint.js (hunk 2 above)
Comment (maxammann) [reply src/Endpoint.js:67]: If the FetchError is never triggered then the FetchError should be removed probably.

Comment (steffenkleinle) [reply src/Endpoint.js:67]: A `FetchError` is thrown in `fetchOrPost` if an error occurs.

File: src/Endpoint.js (hunk 3 above)
Comment (steffenkleinle) [reply src/Endpoint.js]: Renamed. Right now, no different parameters are needed, so I'd leave it as it is. Less duplicated code that way.

File: src/errors/ResponseError.js [hunk 4]
```diff
@@ -1,20 +1,75 @@
...
+    this._url = url
+    this._formData = formData
+  }
+
+  createMessage ({ formData, url, endpointName, response }: ResponseErrorParamsType): string {
+    const stringifyFormData = (formData: ?FormData) => {
+      if (formData) {
+        const entries = {}
+        for (const entry of formData.entries()) {
+          entries[entry[0]] = entry[1]
```
Comment (sarahsporck) [inline src/errors/ResponseError.js]: do we ever fetch 'multipart/form-data' (files), because in this case it would be good to liit the lenght of the entries.

Comment (steffenkleinle) [reply src/errors/ResponseError.js]: Not that I know of, no. Actually, form-data is just used to post feedback.

**Comment by maxammann** [comment]
I can not see a corresponding task in the react native app. Is this only a bugfix for the APICLIENT? If yes, then this should be renamed/moved to the APICLIENT.

File: src/errors/ResponseError.js [hunk 5]
```diff
@@ -1,20 +1,75 @@
...
+        }
+
+        return ` and the formData ${JSON.stringify(entries)}`
+      }
+      return ''
+    }
+
+    return `ResponseError:
+     Failed to load the request for the ${endpointName} endpoint with the url ${url}${stringifyFormData(formData)}.
```
Comment (sarahsporck) [inline src/errors/ResponseError.js]: Maybe you could also add the request method to the message. I think its easier this way to understand that you would expect form data content or an encoded query string, depending on the situation.
```suggestion
     Failed to ${request.method} the request for the ${endpointName} endpoint with the url ${url}${stringifyFormData(formData)}.
```

**Comment by maxammann** [comment]
As this is a "core-functionality" we should test the code of Endpoint.js. Should we do it in this task or create a separate one?

**Comment by lulo4907** [comment]
The code looks fine. But I would like to have some input to reproduce the error and the software change. :)

**Comment by steffenkleinle** [comment]
@maxammann here is the issue: https://issues.integreat-app.de/browse/NATIVE-500. There is already an issue to test the endpoint: https://issues.integreat-app.de/browse/APICLIENT-21

**Comment by steffenkleinle** [comment]
@lulo4907 as mentioned in the description, the aim of this PR is not to solve the issue (since I was not able to reproduce it), but instead it should provide more information on how to reproduce it by adding more information to the error message (which is sent to sentry).

**Review by sarahsporck — APPROVED** [review]
So in regard of the code this pull request looks fine. However I feel we should match it to a separate issue, as the issue is obviously not resolved, but posponed by this PR.

**Comment by steffenkleinle** [comment]
@Taggotty I'll just create a new issue as it is easier to rename the issue than to rename the branch and all commits. (https://issues.integreat-app.de/browse/NATIVE-525)

File: src/errors/ResponseError.js [hunk 6]
```diff
@@ -1,20 +1,75 @@
...
+    this._endpointName = endpointName
+    this._response = response
+    this._url = url
+    this._requestOptions = requestOptions
+  }
+
+  createMessage ({ requestOptions, url, endpointName, response }: ResponseErrorParamsType): string {
+    const stringifyFormData = (formData: FormData) => {
+      const entries = {}
+      for (const entry of formData.entries()) {
```
Comment (maxammann) [inline src/errors/ResponseError.js]: Minor:

```suggestion
      for (const [key, value] of formData.entries()) {
```

**Review by maxammann — APPROVED** [review]

File: src/errors/ResponseError.js [hunk 7]
```diff
@@ -1,20 +1,75 @@
...
+        entries[entry[0]] = entry[1]
+      }
+
+      return ` and the formData ${JSON.stringify(entries)}`
+    }
+
+    const stringifiedFormData = requestOptions.method === 'POST' ? stringifyFormData(requestOptions.body) : ''
+
+    return `ResponseError: Failed to ${requestOptions.method} the request for the ${endpointName} endpoint with the url
+     ${url}${stringifiedFormData}. Received response status ${response.status}: ${response.statusText}.`
```
Comment (maxammann) [inline src/errors/ResponseError.js:48]: I think this error message will look weird as a stringified json object will be appended to an `url`

Comment (steffenkleinle) [reply src/errors/ResponseError.js:48]: Have a look on the code above, I add a string in between: https://github.com/Integreat/integreat-api-client/pull/31/files/f961bab0f5af06cd6c3fdaeb51b2f20c51b1e24b#diff-29f4abd94ac4c57fa229c029b3ea7e6eR42
